### PR TITLE
[develop][REVIEW] Trello-MvfsIe31: Add support of ES6

### PR DESCRIPTION
**Pull request by AndreySurzhan**
[Trello-MvfsIe31](https://trello.com/c/MvfsIe31): Add support of ES6
- [x] Add babel to the project
- [x] Possible to transcompiling ES6 code to ES5 and build prod version of the app
- [x] Update README.md
- [x] Change `var` to `let` in whole project

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -67,9 +67,25 @@ MongoDB:
...
 ```
 App:
+
+To launch develop version of the server:
+
+```
+$ npm run dev
+```
+
+To build production version of the application (it will be build in `./build` directory):
```
Comment (AndreySurzhan) [inline README.md]: built

File: README.md [hunk 2]
```diff
@@ -67,9 +67,25 @@ MongoDB:
...
+
 ```
-$ nodemon server.js
+$ npm run build
 ```
+
+To launch production version of the server (NOTE: prod version should be build for it):
```
Comment (AndreySurzhan) [inline README.md]: built

File: package.json [hunk 3]
```diff
@@ -37,9 +40,16 @@
...
+        "build": "npm run clean && babel . -d build --copy-files --ignore node_module",
+        "prod": "nodemon build/server.js",
+        "lint": "jshint .",
+        "validate": "npm ls"
     },
     "pre-commit": [
         "jshint"
     ]
 }
+
```
Comment (AndreySurzhan) [inline package.json]: empty line

**Comment by AndreySurzhan** [comment]
NOTE:
if you launch npm run build -> all files that is in root directory (except node_modules) will be transcompiled to /build directory.
My concern is  do we need to ignore some files like: gulp files, .gitignore and etc.?

**Comment by AndreySurzhan** [comment]
@wildeastengineer @dashaezhova please review

**Comment by AndreySurzhan** [comment]
NOTE: I change structure a bit
right now we have all source files in `src` folder and `build` with transpiled files that we will be using for prod

**Comment by dimkahare** [comment]
@AndreySurzhan I think variables for modules should be initialized as `const`

**Comment by AndreySurzhan** [comment]
@wildeastengineer @dashaezhova please review
I made all build via gulp. So right now we can have working ES5 project in ./build directory with all non-dev modules.

File: gulpfile.babel.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+const gulp = require('gulp');
+const nodemon = require('gulp-nodemon');
+const dependencies = require('gulp-npm-files');
+const jshint = require('gulp-jshint');
+const jshintConfig = require('./jshintConfig');
+const rimraf = require('rimraf');
+const runSequence = require('run-sequence');
+const shell = require('gulp-shell');
+// Local variables
+let filesToJsHint;
```
Comment (wildeastengineer) [inline gulpfile.babel.js]: lets declare `const`s before `let`s

File: gulpfile.babel.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    'models/**/*.js',
+    'routes/*.js',
+    'utils/*.js',
+    'utils/**/*.js',
+    'gulpfile.babel.js',
+    'jshintConfig.js',
+    'server.js'
+].map(function (file) {
+        return `${sourceFolder}/${file}`;
+    });
```
Comment (wildeastengineer) [inline gulpfile.babel.js]: extra indent

File: gulpfile.babel.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
+// Libraries
```
Comment (wildeastengineer) [inline gulpfile.babel.js]: Why did you rename this file?

File: gulpfile.babel.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        return `${sourceFolder}/${file}`;
+    });
+
+foldersToJsDoc = [
+    'controllers',
+    'models',
+    'routes'
+].map(function (folder) {
+        return `${sourceFolder}/${folder}`;
+    });
```
Comment (wildeastengineer) [inline gulpfile.babel.js]: extra indent

File: gulpfile.babel.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    stream.on('restart', function () {
+        console.log('======== restarted! ======== ')
+    })
+        .on('crash', function() {
+            console.error('======== Application has crashed! ======== \n');
+            stream.emit('restart', 10);  // restart the server in 10 seconds
+        })
+});
+
+/// Start build tasks
```
Comment (wildeastengineer) [inline gulpfile.babel.js]: If there is a lot of tasks and you need to group them, maybe it would be better to exclude the to separate file?

**Comment by wildeastengineer** [comment]
@AndreySurzhan could you make 3 different PR:
1. Integrating es6 support
2. Replacing `var` and etc
3. Docs regenerating

2 and 3 can be in one PR but in different commits.

Thanks!

File: gulpfile.babel.js (hunk 6 above)
Comment (AndreySurzhan) [inline gulpfile.babel.js]: We then need to instruct gulp to use Babel. To do this, we need to rename the gulpfile.js to gulpfile.babel.js

https://markgoodyear.com/2015/06/using-es6-with-gulp/

**Comment by AndreySurzhan** [comment]
@wildeastengineer 
I removed updating commit with documentation from that PR

>@AndreySurzhan could you make 3 different PR:
1. Integrating es6 support
2. Replacing var and etc
3. Docs regenerating

> 2 and 3 can be in one PR but in different commits.

I wouldn't agree on moving replacment var with let to different PR since it is scope of "es6 support" task
I did replacement in separate commit like you did it in web repo

**Comment by AndreySurzhan** [comment]
@wildeastengineer please review

**Comment by wildeastengineer** [comment]
@AndreySurzhan 

> I wouldn't agree on moving replacment var with let to different PR since it is scope of "es6 support" task
I did replacement in separate commit like you did it in web repo

Supporting is not using but I'm ok with the variant with separate commit

File: README.md [hunk 9]
```diff
@@ -67,9 +67,25 @@ MongoDB:
...
+
+```
+$ gulp
+```
+
+To build production version of the application (it will be built in `./build` directory):
+
 ```
-$ nodemon server.js
+$ gulp build:build
```
Comment (wildeastengineer) [inline README.md]: Maybe use `gulp build:launch`?  `build:build` is a kind of tautology.

File: README.md [hunk 10]
```diff
@@ -67,9 +67,25 @@ MongoDB:
 $ sudo service mongod start
 ```
 App:
+
+To launch develop version of the server:
+
+```
+$ gulp
```
Comment (wildeastengineer) [inline README.md]: I would like to have default gulp task that is the only one I have to run on server machine. So lets `gulp` does  build production version of application and run it.
What do you think?

Comment (AndreySurzhan) [inline README.md]: well we are going to have to working versions of the server: dev with es6 code and prod transpired versions in /build directory. When we deploy to prod we could only use code from /build and lunch regular nodemon not uploading the whole repo. So gulp might not be involved in launching prod version at all. Besides as I take it right in davinci you are using grunt default to build dev version.

File: README.md (hunk 9 above)
Comment (AndreySurzhan) [inline README.md]: renamed it to build:transpile since it is more accurate name for that task (convert es6 code to es5)

**Comment by AndreySurzhan** [comment]
@wildeastengineer done, review please
